rotix.1
.\" man page for Rotix
.\" (C) 2001 Sjoerd Hemminga
.\"
.\"	This program is free software; you can redistribute it and/or modify
.\"	it under the terms of the GNU General Public License as published by
.\"	the Free Software Foundation; either version 2 of the License, or
.\"	(at your option) any later version.
.\"
.\"	This program is distributed in the hope that it will be useful, but
.\"	WITHOUT ANY WARRANTY; without even the implied warranty of
.\"	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
.\"	General Public License for more details.
.\"
.\"	You should have received a copy of the GNU General Public License
.\"	along with this program; if not, write to the Free Software
.\"	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
.\"	USA
.TH ROTIX 1 "January 18, 2002" Rotix "Linux User Manuals"
.SH NAME
rotix \- A program to generate rotational obfuscations
.SH SYNOPSIS
.BI "rotix [ \-f " "file " "] [ \-o " "file " "] [ \-r " "value " "]"

.BI "rotix \-t " "text " "[ \-o " "file " "] [ \-r " "value " "]"
.SH DESCRIPTION
.B rotix
is a program that creates rotational obfuscations. Most famously in this
field is the ROT-13 obfuscation.

Please note that
.B rotix
creates obfuscations and not encryptions. You should not consider your data
safe when you ran
.B rotix
over it.
.BR rotix -like
programs are only used to have a "public secret". Mostly to prevent
people from directly reading it, in case of offensive material or the like.
Use it as such and not as an encryption tool. If you want to encrypt your
data have a look at
.BR gpg (1).
.SH OPTIONS
.TP
.B \-f file, \-\-file=file
Specifies a file that contains the text to rotate. This can be stdin
(indicated with a dash "\-").

This option cannot be used in conjunction with
.BR "\-t" ". When neither is given, " "\-f \- " "is assumed."
.TP
.B \-o file, \-\-output=file
The output is appended to the specified file. You can specify stdout (again,
indicated by a single dash "\-"). In fact, stdout is assumed when this
option is omitted.
.TP
.B \-r value, \-\-rot=value
Specifies the rotational value to use. When
.BR "\-r " "is omitted, " "\-r 13 "
is assumed. Negative values are accepted.

A special value is 'auto'.
.B rotix
will then output every unique obfuscation. Please note that this generates
at least 25 lines of output. You might want to specify an output-file or
pipe it to more or less.
.TP
.B \-a, \-\-auto
Alias for
.B \-r auto
.TP
.B \-t text, \-\-text=text
This specifies the text to rotate. You cannot use it in conjunction with
.BR \-f .

Please note that
.B \-t
takes exactly 1 (one) parameter. So if you have special characters (like
spaces) in your text, you might want to escape or quote them.
.TP
.B \-R, \-\-right
This rotates the letters to the right. [Default]
.TP
.B \-L, \-\-left
This rotates the letters to the left.
.TP
.B \-h, \-\-help
Displays the command-line options.
.TP
.B \-v, \-\-version
Outputs the version number of
.BR rotix .
.SH TIPS
If you have large amounts of text that is rotated and of which you do not
what the rotational value is, specify the
.B \-t
option and give the first line of the text as parameter. Use
.B \-a
to check which rotational value is used. Then use that value to deobfuscate
the entire text. This is easier and faster.

If you specify
.B \-a
it is always a good idea to use less.

.B Rotix
has the ability to recalculate values. A rotational value of 27 will be
recalculated to 1.
.B Rotix
can also handle negative values. However, you can use the
.BR \-L " and " \-R " options"
as well to keep using the same values. It is a matter of taste.
.SH COPYRIGHT
Rotix is Copyright \(co 2001 Sjoerd Hemminga

This is free software; see the source for copying conditions. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH REPORTING BUGS
Please report bugs to <sjoerd@huiswerkservice.nl>
.SH AUTHOR
Sjoerd Hemminga <sjoerd@huiswerkservice.nl>
.SH SEE ALSO
.BR gpg (1)


CHANGES
03 Apr 2002: Rotix 0.83

- Fixed a bug in the Makefile, which caused problems when compiling for
  FreeBSD
- Actually made -R/-L working. They were still bound to old test-switches
  (-d/-D).
- The ./configure now recognizes The Hurd and rotix has been reported to
  work.

06 Feb 2002: Rotix 0.82

- Fixed a bug in automated stdin-obfuscation.

27 Jan 2002: Rotix 0.81

- Added an option -a as alias for -r auto.
- Added a ./configure Thanks to Wilmer van der Gaast.
- Added some patches to increase portability.
- Added the possibility to automate obfuscation of stdin.

10 Nov 2001: Rotix 0.8

- Added command-line options that allows the user to control the direction
  of rotating (-R and -L).
- Added I18N support.
- Changed the --help output, so it fits on one screen.
- Fixed a bug, which allowed both stdin and a text to be specified on the
  command-line, when entered in a certain sequence.

13 Oct 2001: Rotix 0.71
- Fixed a bug, which causes blank lines to have the contents of the previous
  line.

18 Sep 2001: Rotix 0.7
- First release.


rotix.c
/*\
 *	Rotix - A program to generate rotational obfuscations
 *	Copyright (C) 2001  Sjoerd Hemminga <sjoerd@huiswerkservice.nl>
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; either version 2 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful, but
 *	WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *	General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program; if not, write to the Free Software
 *	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 *	USA
\*/


/* Rotix includes. */
#include "rot.h"
#include "help.h"

/* System includes. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef NO_GETOPT_LONG
#include <unistd.h>
#define getopt_long(a, b, c, d, e) getopt(a, b, c)
#else
#include <getopt.h>
#endif

/* I18N */
#ifdef I18N
#include <libintl.h>
#define _(String) gettext (String)
#else
#define _(String) String
#endif

/* Defining the flags. */
#define ROT_AUTO		1
#define ROT_INPUT_FILE		2
#define ROT_TEXT		4
#define ROT_OUTPUT_FILE		8
#define ROT_STDIN		16

int main (int argc, char **argv)
{
	int i = 1, rotor = 13, option = 0;
	char *arg = NULL, option_flags = 0, rotate_flags = 0;
	
#ifndef NO_GETOPT_LONG
	/* The struct used to determine which options are called. */
	static struct option rotix_options[] = {
		{ "file",	1,	NULL,	'f' },
		{ "rot",	1,	NULL,	'r' },
		{ "auto",	0,	NULL,	'a' },
		{ "output",	1,	NULL,	'o' },
		{ "text",	1,	NULL,	't' },
		{ "left",	0,	NULL,	'L' },
		{ "right",	0,	NULL,	'R' },
		{ "help",	0,	NULL,	'h' },
		{ "version",	0,	NULL,	'v' },
		{ NULL,		0,	NULL,	 0  }
	};
#endif


	/* A 2-bit character array, to obfuscate a character */
	char rotme[2];
	
	FILE *ifp = NULL;
	FILE *ofp = NULL;

#ifdef I18N
	/* Initialize I18n */
	setlocale (LC_MESSAGES, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);
#endif

	/* Initialization of some variables */	
	rotme[1] = '\0';
	rotate_flags |= ROTATE_RIGHT_FLAG;

/* Parsing options. */
	while (1) {
		option = getopt_long (argc, argv, "f:r:ao:t:LRhv", rotix_options, NULL);
		if (option == -1) break;
		
		switch (option) {
			case '?':
				display_help ();
				exit (1);
			case ':':

			/* Missing argument. */
				display_help ();
				exit (1);
			case 'v':
				display_version ();
				exit (0);
			case 'h':
				display_help ();
				exit (0);

			/* Options that affect the running of Rotix. */
			
			/* Direction right */
			case 'R':
				rotate_flags |= ROTATE_RIGHT_FLAG;
			break;
			
			case 'L':
				if (rotate_flags & ROTATE_RIGHT_FLAG) {
					rotate_flags -= 1;
				}
			break;

			/* Input file */
			case 'f':
				if (strcmp (optarg, "-") == 0) {
					ifp = stdin;
					option_flags |= ROT_STDIN;
					
					if (option_flags & ROT_TEXT) {
						fprintf (stderr, _("rotix: you cannot specify both an inputfile and a text\n"));
						exit (1);
					}
				} else {
					ifp = fopen (optarg, "r");
					option_flags |= ROT_INPUT_FILE;

					if (ifp == NULL) {
						fprintf (stderr, _("rotix: error opening file %s for reading\n"), optarg);
						exit (1);
					} else if (option_flags & ROT_TEXT) {
						fprintf (stderr, _("rotix: you cannot specify both an inputfile and a text\n"));
						exit (1);
					}
				}
			break;

			/* Rotate number */
			case 'r':
				if (strcmp (optarg, "auto") == 0) {
					option_flags |= ROT_AUTO;
				} else if ( !sscanf (optarg, "%9i", &rotor) ) {
					display_help ();
					exit (1);
				}
			break;
			
			/* Automated rotating */
			case 'a':
				option_flags |= ROT_AUTO;
			break;

			/* Output file */
			case 'o':
				if (strcmp (optarg, "-") == 0) {
					ofp = stdout;
				} else {
					ofp = fopen (optarg, "a");

					option_flags |= ROT_OUTPUT_FILE;
					
					if (ofp == NULL) {
						fprintf (stderr, _("rotix: error opening file %s for writing\n"), optarg);
						exit (1);
					}
				}
			break;

			/* Text */
			case 't':
				if ((option_flags & ROT_INPUT_FILE) || (option_flags & ROT_STDIN)) {
					fprintf (stderr, _("rotix: you cannot specify both an inputfile and a text\n"));
					exit (1);
				}
				
				arg = (char *)malloc (strlen (optarg) + 1);
				if (arg == NULL) {
					fprintf (stderr, _("rotix: not enough memory\n"));
					exit (1);
				}
				strcpy (arg, optarg);
				option_flags |= ROT_TEXT;
			break;
			
			/* Unrecognized option. NOTE: this should never happen, since case '?' should catch this one */
			default:
				fprintf (stderr, _("rotix: fixme\n"));
				exit (2);
		}
	}

	if (!(option_flags & ROT_INPUT_FILE) && !(option_flags & ROT_TEXT)) {
		option_flags |= ROT_STDIN;
		ifp = stdin;
	}
	
	if (ofp == NULL) {
		ofp = stdout;
	}
	
/*\
 *  End of option-parsing.
\*/
	if (option_flags & ROT_TEXT) {
		if (option_flags & ROT_AUTO) {
			for (i = 1; i < 26; i++) {
				rotate (1, arg, rotate_flags);
				fprintf (ofp, "ROT-%02i: %s\n", i, arg);
			}
			
			free (arg);
		} else {
			rotate (rotor, arg, rotate_flags);
			fprintf (ofp, "%s\n", arg);
		}
	} else if (option_flags & ROT_AUTO) {
		/* Goodie, goodie, automated (de-)obfuscation of a file! */
		
		if (ifp == stdin) {
			/* Very dirty hack to get this functioning without using any extra variables */
			ifp = tmpfile ();
			
			if (ifp == NULL) {
				fprintf (stderr, _("rotix: error creating temporary file\n"));
				exit (1);
			}
			
			arg = (char *)malloc (265);
			if (arg == NULL) {
				fprintf (stderr, _("rotix: not enough memory\n"));
				exit (1);
			}
			
			while ( !feof (stdin) ) {
				strcpy (arg, "");
				rotme[0] = '\0';
				
				fscanf (stdin, "%255[^\n]s", arg);
				fprintf (ifp, "%s", arg);
				
				if (feof (stdin)) break;
				
				rotme[0] = fgetc (stdin);
				fprintf (ifp, "%s", rotme);
			}
			rewind (ifp);
		}
		
		arg = (char *)malloc (256);
		if (arg == NULL) {
			fprintf (stderr, _("rotix: not enough memory\n"));
			exit (1);
		}

		for (i = 1; i < 26; i++) {
			fprintf (ofp, "\nROT-%02i:\n=======\n\n", i);
			
			while ( !feof (ifp) ) {
				strcpy (arg, "");
				rotme[0] = '\0';

				fscanf (ifp, "%255[^\n]s", arg);
				rotate (i, arg, rotate_flags);
				fprintf (ofp, "%s", arg);

				if (feof (ifp)) break;

				rotme[0] = fgetc (ifp);
				rotate (i, rotme, rotate_flags);
				fprintf (ofp, "%s", rotme);
			}
			rewind (ifp);
		}
		free (arg);
	} else {
		arg = (char *)malloc (256);
		if (arg == NULL) {
			fprintf (stderr, _("rotix: not enough memory\n"));
			exit (1);
		}

		while ( !feof (ifp) ) {
			strcpy (arg, "");
			rotme[0] = '\0';

			fscanf (ifp, "%255[^\n]s", arg);
			rotate (rotor, arg, rotate_flags);
			fprintf (ofp, "%s", arg);

			if (feof (ifp)) break;

			rotme[0] = fgetc (ifp);
			rotate (rotor, rotme, rotate_flags);
			fprintf (ofp, "%s", rotme);
		}

		free (arg);
	}

	fflush (NULL);

	if (option_flags & ROT_INPUT_FILE) {
		fclose (ifp);
	}
	
	if (option_flags & ROT_OUTPUT_FILE) {
		fclose (ofp);
	}		
	
	return (0);
}
